[WIP] Fix search() when mount point overlaps an existing folder

PVince81 opened the pull request:
Currently it's still possible to make a system mount point overlap user folders, so we need a workaround to avoid searching into these hidden/overlapped folders.

In the spirit of "test first", this PR first provides unit tests to cover this case.
To make the matter worse: there are two places in the code that implement a similar logic (duplicate code!)

Fix for https://github.com/owncloud/core/issues/13096#issuecomment-70246172
- [x] actual fix
- [x] unit test

To fix this I'll probably have to iterate over the search results and delete the ones that match an existing mount point. Might not be good for performance.

CC @MorrisJobke @icewind1991

Commit:
Add unit test for overlapping mount + folder

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c08cfe3a-b384-4e2d-8e30-4ac0ef568bc1) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c08cfe3a-b384-4e2d-8e30-4ac0ef568bc1) completed: **4 new issues, 2 updated code elements**

karlitschek commented:
:+1:

Commit:
Prevent searching in folders hidden by mounts

Note: this fix is only for mount points hiding folders that are in the
root storage.

This does not fix the case where one secondary storage would be mounted
on top of another third storage...

PVince81 commented:
I give up trying to refactor Node API + View's search into one (WIP here: https://github.com/owncloud/core/pull/13420)

This means the fix will need to be put in two separate places

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/bb643d41-df38-48e4-9def-d6dd8d3782e8) was created.

PVince81 commented:
Here we go, I added an extra check in both locations. It is likely to kill performance a bit when searching through big folders because for every result we need to check whether it's from a hidden folder.

I'm not sure whether fixing this is even a good idea... ideally it shouldn't be possible to cover existing folders when mounting storages.

@icewind1991 please have a look and tell me this is a bad fix and that it indeed kills performance...
Another idea would be to pass a list of excluded folder to the cache's "search()" methods, and exclude using a `WHERE` clause. But it all sounds like overkill.

PVince81 commented:
I haven't tested this extensively yet because I'd like to discuss the solution first.
(but it does fix https://github.com/owncloud/core/issues/13096#issuecomment-70246172)

icewind1991 commented on lib/private/files/node/folder.php:
```diff
@@ -264,30 +264,46 @@ private function searchCommon($method, $args) {
-					$result['storage'] = $storage;
-					$files[] = $result;
-				}
+		foreach ($mountPoints as list($subStorage, $relativeMountPoint)) {
```
`as list(...)` only work in php5.5+

PVince81 replied:
Ahh.... for once that I use fancy stuff :wink:

PVince81 commented:
Also note that this doesn't cover the crazy cases where ext storage 1 is mounted onto a subfolder of ext storage 2... that would be even worse for performance.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/bb643d41-df38-48e4-9def-d6dd8d3782e8) completed: **19 new issues, 4 updated code elements**

PVince81 commented:
I don't like this approach, introduces performance issues to only fix a corner case.

Closing.
